config/configcheck/configcheck.go

```diff
@@ -0,0 +1,194 @@
+	}
+	for _, factory := range factories.Extensions {
+		configs = append(configs, factory.CreateDefaultConfig())
+	}
+
+	for _, config := range configs {
+		if err := ValidateConfig(config); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return oterr.CombineErrors(errs)
+}
+
+// ValidateConfig enforces that given configuration object is following the patterns
+// used by the collector. This ensures consistency between different implementations
+// of components and extensions. It is recommended for implementers of components
+// to call this function on their tests passing the default configuration of the
+// component factory.
+func ValidateConfig(config interface{}) error {
+	t := reflect.TypeOf(config)
+
+	tk := t.Kind()
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		tk = t.Kind()
+	}
+
+	if tk != reflect.Struct {
+		return fmt.Errorf(
+			"config must be a struct or a pointer to one, the passed object is a %s",
+			tk)
+	}
+
+	return validateConfigDataType(t)
+}
+
+// validateConfigDataType performs a descending validation of the given type.
+// If the type is a struct it goes to each of its fields to check for the proper
+// tags.
+func validateConfigDataType(t reflect.Type) error {
+	var errs []error
+
+	switch t.Kind() {
+	case reflect.Ptr:
+		if err := validateConfigDataType(t.Elem()); err != nil {
+			errs = append(errs, err)
+		}
+	case reflect.Struct:
+		// Reflect on the pointed data and check each of its fields.
+		// vt := reflect.ValueOf(config).Elem().Type()
+		nf := t.NumField()
+		for i := 0; i < nf; i++ {
+			f := t.Field(i)
+			if err := checkStructFieldTags(f); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	default:
+		// The config object can carry other types but they are not used when
+		// reading the configuration via viper so ignore them. Basically ignore:
+		// reflect.Uintptr, reflect.Chan, reflect.Func, reflect.Interface, and
+		// reflect.UnsafePointer.
+	}
+
+	if err := oterr.CombineErrors(errs); err != nil {
+		return fmt.Errorf(
+			"type %q from package %q has invalid config settings: %v",
+			t.Name(),
+			t.PkgPath(),
+			err)
+	}
+
+	return nil
+}
+
+// checkStructFieldTags inspects the tags of a struct field.
+func checkStructFieldTags(f reflect.StructField) error {
+
+	tagValue := f.Tag.Get("mapstructure")
+	if tagValue == "" {
+
+		// Ignore special types.
+		switch f.Type.Kind() {
+		case reflect.Interface, reflect.Chan, reflect.Func, reflect.Uintptr, reflect.UnsafePointer:
+			// Allow the config to carry the types above, but since they are not read
+			// when loading configuration, just ignore them.
+			return nil
+		}
+
+		// Public fields of other types should be tagged.
+		chars := []byte(f.Name)
+		if len(chars) > 0 && chars[0] >= 'A' && chars[0] <= 'Z' {
+			return fmt.Errorf("mapstructure tag not present on field %q", f.Name)
+		}
+
+		// Not public field, no need to have a tag.
+		return nil
+	}
+
+	tagParts := strings.Split(tagValue, ",")
```

Comment:
I am a bit worried that we parse the tags ourselves. Isn't there a built-in Go lib that we can ask to do this for us?

Reply:
Unfortunately nothing in standard lib, the value itself is up to each key/tag.